### PR TITLE
f-checkout@3.13.0 - Split notes v2

### PR DESCRIPTION
**Pull request by charliearlie**
Split notes work for Checkout Web (awaiting backend changes)

### CODE REVIEW

File: packages/components/pages/f-checkout/src/checkout-issues.js [hunk 1]
```diff
@@ -41,6 +41,21 @@ const checkoutIssues = {
         shouldRedirectToMenu: false
     },
 
+    KITCHEN_NOTE_NOT_ACCEPTED: {
+        shouldShowInDialog: true,
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/checkout-issues.js]: should now be `errorType: ERROR_TYPES.dialog` (same below)

File: packages/components/pages/f-checkout/src/components/Checkout.vue [hunk 2]
```diff
@@ -832,17 +849,39 @@ export default {
...
+
+        async loadNotesConfiguration () {
+            if (this.checkoutFeatures.isSplitNotesEnabled) {
+                try {
+                    await this.getNotesConfiguration({
+                        url: `${this.getNoteConfigUrl}/${this.restaurant.id}/checkout-note-types`,
+                        timeout: this.checkoutTimeout
+                    });
+                } catch (error) {
+                    this.logInvoker({
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/components/Checkout.vue]: could you use `handleEventLogging` here?

File: packages/components/pages/f-checkout/src/components/Notes.vue [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    <div v-if="isSplitNotesEnabled">
+        <accordion
+            :id="noteTypeCourierOrOrder"
+            :title="$t(`userNote.${noteTypeCourierOrOrder}.${serviceType}.title`)">
+            <span :class="$style['c-checkout-accordion-help']">{{ $t(`userNote.${noteTypeCourierOrOrder}.${serviceType}.text`) }}</span>
+            <form-field
+                input-type="textarea"
+                :placeholder="$t(`userNote.${noteTypeCourierOrOrder}.${serviceType}.placeholder`)"
+                :value="noteValue"
+                cols="30"
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/components/Notes.vue]: could you have `cols`, `rows`, `maxlength` and any other prop thats reused here as a computed and then `v-bind="textAreaProps"`

**Review by FayeCarter — DISMISSED** [review]

**Comment by matthewhardern** [comment]
Please can you update from master to join the build approval process, thanks

**Review by FayeCarter — DISMISSED** [review]

**Review by FayeCarter — DISMISSED** [review]

File: .husky/pre-commit [hunk 4]
```diff
@@ -1,5 +0,0 @@
-#!/bin/sh
-. "$(dirname "$0")/_/husky.sh"
```
Comment (xander-marjoram) [inline .husky/pre-commit]: Is this supposed to be removed?

Comment (charliearlie) [reply .husky/pre-commit]: Nope. Good spot. Not sure how that happened. 😅

File: packages/components/pages/f-checkout/src/components/Checkout.vue [hunk 5]
```diff
@@ -332,6 +346,16 @@ export default {
...
             };
+        },
+
+        placeOrderNotes () {
+            const { courier, kitchen, order } = this.notes;
+            return {
+                NoteForRestaurant: order?.note || null,
+                ...(kitchen?.note && { NoteForKitchen: kitchen.note }),
+                ...(courier?.note && { NoteForDelivery: courier.note }),
+                ...(courier?.note && { NoteForDriver: courier.note })
```
Comment (charliearlie) [inline packages/components/pages/f-checkout/src/components/Checkout.vue]: Unfortunately, yes we do need to add the same field twice under 2 different keys 🤦‍♂️

Comment (DamianMullins) [reply packages/components/pages/f-checkout/src/components/Checkout.vue]: You could reduce this down a little?

```js
...(courier?.note && { 
	NoteForDelivery: courier.note,
	NoteForDriver: courier.note
})
```

Comment (charliearlie) [reply packages/components/pages/f-checkout/src/components/Checkout.vue]: Much better. Thanks @DamianMullins - it was a last minute requirement so I just slapped another property on.

**Review by FayeCarter — DISMISSED** [review]

File: packages/components/pages/f-checkout/src/components/Accordion.vue [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            this.isExpanded = !this.isExpanded;
+        }
+    }
+};
+</script>
+
+<style lang="scss" module>
+    $icon-size: 16px;
+    .c-checkout-accordion-header {
+        font-size: 1rem;
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/Accordion.vue]: Are there fozzie variables for these?

File: packages/components/pages/f-checkout/src/components/Checkout.vue [hunk 7]
```diff
@@ -832,17 +852,35 @@ export default {
...
+                        url: `${this.getNoteConfigUrl}/${this.restaurant.id}/checkout-note-types`,
+                        timeout: this.checkoutTimeout
+                    });
+                } catch (error) {
+                    this.handleEventLogging('NotesConfigurationFailure');
+                }
+            }
+        },
+
+        getNotes () {
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/Checkout.vue]: Does this method belong in the store?

Comment (xander-marjoram) [reply packages/components/pages/f-checkout/src/components/Checkout.vue]: And maybe `loadNotesConfiguration`

File: packages/components/pages/f-checkout/src/components/Notes.vue [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            <form-field
+                :placeholder="$t(`userNote.kitchen.${serviceType}.placeholder`)"
+                :value="kitchenNoteValue"
+                v-bind="inputStyles"
+                name="kitchen-note"
+                @input="updateUserNotes({ note: $event, type: 'kitchen' })" />
+        </accordion>
+    </div>
+    <form-field
+        v-else-if="!isSplitNotesEnabled"
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/Notes.vue]: Would simply `v-else` work here or does the `v-if` on L15 interfere?

File: packages/components/pages/f-checkout/src/components/Notes.vue [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        v-bind="inputStyles"
+        :name="'order-note'"
+        @input="updateUserNotes({ note: $event, type: 'order' })" />
+</template>
+
+<script>
+import { mapActions, mapState } from 'vuex';
+import FormField from '@justeat/f-form-field';
+import Accordion from './Accordion.vue';
+import '@justeat/f-form-field/dist/f-form-field.css';
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/Notes.vue]: Nit: I think it's good to keep the external and internal imports separate
```suggestion
import FormField from '@justeat/f-form-field';
import '@justeat/f-form-field/dist/f-form-field.css';

import Accordion from './Accordion.vue';
```

File: packages/components/pages/f-checkout/src/components/Notes.vue [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    },
+
+    computed: {
+        ...mapState(VUEX_CHECKOUT_MODULE, [
+            'notesConfiguration',
+            'serviceType',
+            'notes'
+        ]),
+
+        shouldShowKitchenNotes () {
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/Notes.vue]: This could be a getter in the store?

Comment (xander-marjoram) [reply packages/components/pages/f-checkout/src/components/Notes.vue]: And same for lots of the below computed properties, which seem to only depend on values from the store's state

File: packages/components/pages/f-checkout/src/components/Notes.vue [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+        noteValue () {
+            return this.noteTypeCourierOrOrder === CHECKOUT_NOTE_TYPE_COURIER ? this.notes.courier?.note : this.notes.order?.note;
+        },
+
+        kitchenNoteValue () {
+            return this.notes.kitchen?.note || '';
+        },
+
+        inputStyles () {
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/Notes.vue]: Assuming this is the last remaining computed property, it could be moved into `data` seeing as it doesn't need to be recomputed

File: packages/components/pages/f-checkout/CHANGELOG.md [hunk 12]
```diff
@@ -3,6 +3,13 @@
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+v3.13.0
+------------------------------
+*December 10, 2021*
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/CHANGELOG.md]: If you make any other changes, might as well bump the date while you're there!

File: packages/components/pages/f-checkout/src/components/_tests/Accordion.test.js [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                    }
+                });
+
+                // Act
+                wrapper.vm.toggle();
+
+                // Assert
+                expect(wrapper.vm.$data.isExpanded).toBe(false);
+            });
+        });
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/_tests/Accordion.test.js]: You can combine these two tests with `it.each`
```suggestion
            it.each([
                [true, false],
                [false, true]
            ])('should change `isExpanded` to %s when it is initially set to %s', (expected, initial) => {
                // Arrange
                wrapper = shallowMount(Accordion, {
                    i18n,
                    store: createStore(),
                    localVue,
                    propsData,
                    data () {
                        return {
                            isExpanded: initial
                        };
                    }
                });

                // Act
                wrapper.vm.toggle();

                // Assert
                expect(wrapper.vm.$data.isExpanded).toBe(expected);
            });
        });
```

File: packages/components/pages/f-checkout/src/components/_tests/Accordion.test.js [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+        describe('template ::', () => {
+            let wrapper;
+
+            it('should toggle the accordion visibility when heading is clicked', () => {
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/_tests/Accordion.test.js]: Might be worth another test to check that it closes too, and you should be able to use `it.each` again

File: packages/components/pages/f-checkout/src/components/_tests/Checkout.test.js [hunk 15]
```diff
@@ -3133,6 +3135,62 @@ describe('Checkout', () => {
...
+            });
+
+            afterEach(() => {
+                jest.clearAllMocks();
+            });
+
+            it('should call `placeOrder`', async () => {
+                // Arrange
+                const placeOrderSpy = jest.spyOn(wrapper.vm, 'placeOrder');
+                jest.spyOn(wrapper.vm, 'getReferralState').mockImplementation(() => 'MockReferralState');
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/_tests/Checkout.test.js:3169]: I think it would be better to mock the cookie value instead, so that `getReferralState` doesn't need mocking

File: packages/components/pages/f-checkout/src/tenants/en-GB.js [hunk 16]
```diff
@@ -151,6 +151,24 @@ const messages = {
                 buttonText: 'Back to order'
             },
 
+            KITCHEN_NOTE_NOT_ACCEPTED: {
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/tenants/en-GB.js:154]: Just checking, do these error messages mean that there was a problem submitting the note or were the customer's notes actually rejected?

File: packages/components/pages/f-checkout/src/tenants/en-GB.js [hunk 17]
```diff
@@ -182,22 +200,57 @@ const messages = {
...
+        kitchen: {
+            delivery: {
+                title: 'Leave a note for the kitchen',
+                text: 'Leave a note for the restaurant with anything they need to know.',
+                placeholder: 'We\'re working with restaurants to cut waste. Please don\'t ask for plastic cutlery.'
+            },
+            dinein: {
+                title: 'Leave a note for the kitchen',
+                text: 'Leave a note for the restaurant with anything they need to know.',
+                placeholder: 'We\'re working with restaurants to cut waste. Please don\'t ask for plastic cutlery.'
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/tenants/en-GB.js:246]: Asking for plastic cutlery for dine-in sounds interesting 😄

File: packages/components/pages/f-checkout/src/tenants/en-GB.js (hunk 16 above)
Comment (charliearlie) [reply packages/components/pages/f-checkout/src/tenants/en-GB.js:154]: They were rejected because the config changed as they were checking out. It's a miniscule edge case, but this was the agreed copy from product.

File: packages/components/pages/f-checkout/test/visual/f-checkout-authenticated.visual.desktop.spec.js [hunk 18]
```diff
@@ -315,6 +315,17 @@ describe('f-checkout - Dine In - Authenticated - Desktop Visual Tests', () => {
...
         browser.percyScreenshot('f-checkout - Dine in - Authenticated - "Duplicate Order Warning" Modal', 'desktop');
     });
+
+    it('should display the two notes fields if there is two noteTypes.', () => {
+        // Arrange
+        checkout = new Checkout('organism', 'checkout-component');
+        checkout.withQuery('&knob-Service Type', 'collection')
+            .withQuery('&knob-Is User Logged In', true)
+            .withQuery('&knob-Is ASAP available', true)
+            .withQuery('$knob-Note%20types', 'get-notes-config-split');
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/test/visual/f-checkout-authenticated.visual.desktop.spec.js]: I think it handles spaces already
```suggestion
            .withQuery('$knob-Note types', 'get-notes-config-split');
```

File: packages/components/pages/f-checkout/src/components/Accordion.vue (hunk 6 above)
Comment (charliearlie) [reply packages/components/pages/f-checkout/src/components/Accordion.vue]: Yep, and have updated 🙂

File: packages/components/pages/f-checkout/src/components/Checkout.vue (hunk 7 above)
Comment (charliearlie) [reply packages/components/pages/f-checkout/src/components/Checkout.vue]: Moved the `getNotes` into a getter as that makes more sense.

File: packages/components/pages/f-checkout/src/components/Notes.vue (hunk 8 above)
Comment (charliearlie) [reply packages/components/pages/f-checkout/src/components/Notes.vue]: That works, but I guess my fear was that what you said would be the case. Now changed.

File: packages/components/pages/f-checkout/src/components/Notes.vue (hunk 9 above)
Comment (charliearlie) [reply packages/components/pages/f-checkout/src/components/Notes.vue]: Agreed. Oversight on my part.

File: packages/components/pages/f-checkout/src/components/Notes.vue (hunk 10 above)
Comment (charliearlie) [reply packages/components/pages/f-checkout/src/components/Notes.vue]: Yep, I've decided to take this approach with a few of the things you've suggested.

File: packages/components/pages/f-checkout/src/components/_tests/Checkout.test.js (hunk 15 above)
Comment (charliearlie) [reply packages/components/pages/f-checkout/src/components/_tests/Checkout.test.js:3169]: I actually didn't write this test, I just moved it under the describe block 'submitOrder (split notes enabled)::'

File: packages/components/pages/f-checkout/src/components/Notes.vue (hunk 10 above)
Comment (charliearlie) [reply packages/components/pages/f-checkout/src/components/Notes.vue]: Now updated with no computed properties in the notes component!

**Comment by charliearlie** [comment]
Thanks for the review on such a big PR @xander-marjoram - the scope just kept creeping with this one. I have resolved all of the comments you made. 🙂

File: packages/components/pages/f-checkout/src/components/Checkout.vue [hunk 19]
```diff
@@ -332,6 +350,18 @@ export default {
                 [EventNames.FormValid]: this.submitCheckout,
                 [EventNames.FormInvalid]: this.onInvalidCheckoutForm
             };
+        },
+
+        placeOrderNotes () {
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/Checkout.vue:355]: This one could probably come from the store too, but I'm happy if you want to do that in a later PR

File: packages/components/pages/f-checkout/src/components/Notes.vue [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        };
+    },
+
+    computed: {
+        ...mapState(VUEX_CHECKOUT_MODULE, [
+            'features',
+            'notesConfiguration',
+            'serviceType',
+            'notes'
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/Notes.vue]: I think `notesConfiguration` and `notes` are no longer used in this file?
```suggestion
            'serviceType'
```

File: packages/components/pages/f-checkout/src/components/_tests/Accordion.test.js [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+            });
+        });
+
+        describe('template ::', () => {
+            let wrapper;
+
+            it.each([
+                [true, false],
+                [false, true]
+            ])('should toggle the accordion visibility from %s to %s when heading is clicked', (expected, initial) => {
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/_tests/Accordion.test.js]: Nit: The order of the parameters doesn't match how they're ordered in the test description

```suggestion
            ])('should toggle the accordion visibility from %s to %s when heading is clicked', (initial, expected) => {
```

File: packages/components/pages/f-checkout/src/store/tests/checkout.module.test.js [hunk 22]
```diff
@@ -20,7 +20,11 @@ import {
     mockAuthToken, mockAuthTokenNoNumbers, mockAuthTokenNoMobileNumber
 } from '../../components/_tests/helpers/setup';
 import {
-    VUEX_CHECKOUT_ANALYTICS_MODULE, DEFAULT_CHECKOUT_ISSUE, ERROR_TYPES, DUPLICATE_ORDER
+    VUEX_CHECKOUT_ANALYTICS_MODULE,
+    DEFAULT_CHECKOUT_ISSUE,
+    ERROR_TYPES, DUPLICATE_ORDER,
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/store/tests/checkout.module.test.js]: Nit: for consistency
```suggestion
    DUPLICATE_ORDER,
    ERROR_TYPES,
```

Comment (charliearlie) [reply packages/components/pages/f-checkout/src/store/tests/checkout.module.test.js]: For someone with serious OCD around formatting, I'm shocked that I missed this. All done now.

File: packages/components/pages/f-checkout/src/components/Checkout.vue (hunk 19 above)
Comment (charliearlie) [reply packages/components/pages/f-checkout/src/components/Checkout.vue:355]: It's probably going to be a quiet week next week looking at our Jira board, so I will save that for a further PR to have something to do. 😅

File: packages/components/pages/f-checkout/src/components/Notes.vue [hunk 23]
```diff
@@ -69,10 +69,7 @@ export default {
 
     computed: {
         ...mapState(VUEX_CHECKOUT_MODULE, [
-            'features',
-            'notesConfiguration',
-            'serviceType',
```
Comment (xander-marjoram) [inline packages/components/pages/f-checkout/src/components/Notes.vue]: Looks like `serviceType` is still used in the template, right?

Comment (charliearlie) [reply packages/components/pages/f-checkout/src/components/Notes.vue]: Yep, I got too trigger happy and accidentally deleted that as well late on a Friday. 😅

**Review by FayeCarter — DISMISSED** [review]

**Review by xander-marjoram — DISMISSED** [review]

**Review by xander-marjoram — APPROVED** [review]

**Review by FayeCarter — APPROVED** [review]
